testsuite/smoke-test/src/full_nodes.rs

```diff
@@ -269,3 +279,146 @@ fn test_vfn_failover() {
+    let private_swarm = env.public_swarm(PRIVATE);
+    private_swarm.lock().launch();
+
+    // And connect the user to the private swarm
+    add_node_to_seeds(
+        &mut user_config,
+        private_swarm.lock().get_node(0).unwrap().config(),
+        NetworkId::Public,
+        PeerRole::PreferredUpstream,
+    );
+    env.add_public_fn_swarm(USER, 1, Some(user_config), &private_swarm.lock().config);
+    let user_swarm = env.public_swarm(USER);
+
+    user_swarm.lock().launch();
+
+    // Ensure that User node is connected to private node and only the private node
+    {
+        let mut user_swarm = user_swarm.lock();
+        let user_node = user_swarm.nodes.get_mut("0").unwrap();
+        assert_eq!(
+            1,
+            user_node
+                .get_connected_peers(NetworkId::Public, None)
+                .unwrap_or(0),
+            "User node is connected to more than one peer"
+        );
+    }
+
+    // read state from full node client
+    let mut validator_client = env.get_validator_client(0, None);
+    let mut user_client = env.get_pfn_client(USER, 0, None);
+
+    // mint from user node and check both validator and user node have correct balance
+    let _account = validator_client.create_next_account(false).unwrap().address;
+    user_client.create_next_account(false).unwrap();
+
+    let sender_account = testnet_dd_account_address();
+    let creation_account = treasury_compliance_account_address();
+
+    get_and_reset_sequence_number(&mut user_client, sender_account);
+    get_and_reset_sequence_number(&mut user_client, creation_account);
+    mint_coins(&mut user_client, 0, 10, XUS, true);
+
+    assert!(compare_balances(
+        vec![(10.0, XUS.to_string())],
+        get_balances(&mut user_client, 0)
+    ));
+
+    let sequence = get_and_reset_sequence_number(&mut user_client, sender_account);
+    wait_for_transaction(&mut validator_client, sender_account, sequence - 1);
+    assert!(compare_balances(
+        vec![(10.0, XUS.to_string())],
+        get_balances(&mut validator_client, 0),
+    ));
+}
+
+fn add_node_to_seeds(
+    dest_config: &mut NodeConfig,
+    seed_config: &NodeConfig,
+    network_id: NetworkId,
+    peer_role: PeerRole,
+) {
+    let dest_network_config = dest_config
+        .full_node_networks
+        .iter_mut()
+        .find(|network| network.network_id == network_id)
+        .unwrap();
+    let seed_network_config = seed_config
+        .full_node_networks
+        .iter()
+        .find(|network| network.network_id == network_id)
+        .unwrap();
+
+    let seed_peer_id = seed_network_config.peer_id();
+    let seed_key = seed_network_config.identity_key().public_key();
+
+    let seed_peer = if peer_role != PeerRole::Downstream {
+        // For upstreams, we know the address, but so don't duplicate the keys in the config (lazy way)
+        // TODO: This is ridiculous, we need a better way to manipulate these `NetworkAddress`s
```

Comment:
Hah, agreed 😄

Reply:
Ha, I should fix this comment, and editorialize a little less.